crates/matrix-sdk/src/http_client/mod.rs

```diff
@@ -259,3 +295,107 @@ impl tower::Service<http_old::Request<Bytes>> for HttpClient {
         Box::pin(fut)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        http_client::RequestConfig,
+        test_utils::{set_client_session, test_client_builder_with_server},
+    };
+    use matrix_sdk_test::async_test;
+    use std::{
+        num::NonZeroUsize,
+        sync::{
+            atomic::{AtomicU8, Ordering},
+            Arc,
+        },
+        time::Duration,
+    };
+    use wiremock::{matchers::method, Mock, Request, ResponseTemplate};
+
+    #[async_test]
+    async fn ensure_concurrent_request_limit_is_observed() {
+        let (client_builder, server) = test_client_builder_with_server().await;
+        let mut client = client_builder
+            .request_config(RequestConfig::default().max_concurrent_requests(NonZeroUsize::new(5)))
+            .build()
+            .await
+            .unwrap();
+
+        set_client_session(&mut client).await;
+
+        let counter = Arc::new(AtomicU8::new(0));
+        let inner_counter = counter.clone();
+
+        Mock::given(method("GET"))
+            .respond_with(move |_req: &Request| {
+                inner_counter.fetch_add(1, Ordering::SeqCst);
+                // we stall the requests
+                ResponseTemplate::new(200).set_delay(Duration::from_secs(60))
+            })
+            .mount(&server)
+            .await;
+
+        let _bg_task = tokio::spawn(async move {
+            let mut pollers = Vec::new();
+
+            for _n in 0..10 {
+                pollers.push(client.whoami());
+            }
+            // issue parallel execution
+            futures_util::future::join_all(pollers).await
```

Comment:
This might be slightly more concise:

```suggestion
            futures_util::future::join_all((0..10).map(|_| client.whoami()).await
```

Reply:
Yes, with the spawn-move this might work now (had this before but had Lifetime issue for client in that circumstance).